portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,28 +14,108 @@
 import java.util.ArrayList; 
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+/** Servlet responsible for handling comments data */
+@WebServlet("/comment")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> messages = new ArrayList<String>();
-    messages.add("Hello");
-    messages.add("Testing JSON");
-    messages.add("I hope this works!");
+    String sort = request.getParameter("sort");
+
+    // Sort comments according to request parameter
+    Query query;
+    FilterPredicate filter = new FilterPredicate("img", Query.FilterOperator.EQUAL, request.getParameter("img"));
+    if (sort.equals("new")) {
+      query = new Query("Comment").setFilter(filter).addSort("timestamp", SortDirection.DESCENDING);
+    } else {
+      query = new Query("Comment").setFilter(filter).addSort("timestamp", SortDirection.ASCENDING);
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    int numComments = results.countEntities(FetchOptions.Builder.withDefaults());
+
+    // Set the number of comments to print (max)
+    String quantity = request.getParameter("quantity");
+    int max = -1;
+    if (quantity != null && quantity.length() > 0) {
+      max = Integer.parseInt(quantity);
+    } else {
+      max = numComments;
+    }
+
+    // Create Comment objects and add to comments ArrayList
+    ArrayList<Comment> comments = new ArrayList<Comment>();
+    for (Entity entity : results.asIterable(FetchOptions.Builder.withLimit(max))) {
+      long id = entity.getKey().getId();  
+      String comment = (String) entity.getProperty("comment");
+      String name = (String) entity.getProperty("name");
+      long timestamp = (long) entity.getProperty("timestamp");
+      String img = (String) entity.getProperty("img");
+
+      Comment c = new Comment(id, name, comment, timestamp, img);
+      comments.add(c);
+    }
+    
+    // Create json String to print to /comments
     Gson gson = new Gson();
-    String json = gson.toJson(messages);
+    String json = "{";
+    json += "\"comments\": ";
+    json += gson.toJson(comments);
+    json += ", ";
+    json += "\"total\": ";
+    json += numComments;
+    json += "}";
 
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String comment = getParameter(request, "comment", "");
+    String name = getParameter(request, "name", "");
+    String img = getParameter(request, "img", "");
+    long timestamp = System.currentTimeMillis();
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("comment", comment);
```

Comment:
In a design perspective, do you really want to post a comment if the comment is "". Perhaps don't sent create the commentEntity if there is not comment.

Reply:
That's true- I changed it so that I return the function early if it's ""

Reply:
Good point. Another thing is do we still want to add the comment if "img" is ""? Looks like we will never show the comment if (!img.equals(imgRequest)).

Reply:
BTW, these checks should happen on client side (js code) unless we need info on server side. e.g. if comment is empty, we don't send doPost at all (rather than send but early return).

Reply:
Ooo ok I'll move the checks to the client side